File: test_iframes.py
import re
import urllib.request
import ssl

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

with open('src/lib/data/links.ts', 'r') as f:
    content = f.read()

urls = re.findall(r"url:\s*'([^']+)'", content)

for url in urls:
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    try:
        with urllib.request.urlopen(req, context=ctx, timeout=5) as response:
            headers = response.headers
            x_frame = headers.get('X-Frame-Options', '').lower()
            csp = headers.get('Content-Security-Policy', '').lower()
            
            blocks = False
            reason = ''
            if x_frame in ['deny', 'sameorigin']:
                blocks = True
                reason = f"X-Frame-Options: {x_frame}"
            elif 'frame-ancestors' in csp and 'campusweb.cloud' not in csp:
                blocks = True
                reason = f"CSP: {csp}"
                
            print(f"{'BLOCKS' if blocks else 'ALLOWS'} - {url} ({reason})")
    except Exception as e:
        print(f"ERROR - {url} ({e})")

File: test_iframes.js
const fs = require('fs');
const path = require('path');

// A simple regex to find all urls in links.ts
const content = fs.readFileSync('src/lib/data/links.ts', 'utf-8');
const urlRegex = /url:\s*'([^']+)'/g;
let match;
const urls = [];
while ((match = urlRegex.exec(content)) !== null) {
  urls.push(match[1]);
}

async function checkUrl(urlStr) {
  try {
    const res = await fetch(urlStr, { method: 'GET' });
    const xFrameOptions = res.headers.get('x-frame-options');
    const csp = res.headers.get('content-security-policy');
    
    let blocksIframe = false;
    let reason = '';
    
    if (xFrameOptions) {
      if (xFrameOptions.toLowerCase() === 'deny' || xFrameOptions.toLowerCase() === 'sameorigin') {
        blocksIframe = true;
        reason = `X-Frame-Options: ${xFrameOptions}`;
      }
    }
    
    if (csp && csp.toLowerCase().includes('frame-ancestors')) {
      if (!csp.toLowerCase().includes('campusweb.cloud')) {
        blocksIframe = true;
        reason = `CSP: ${csp}`;
      }
    }
    
    console.log(`${blocksIframe ? 'BLOCKS' : 'ALLOWS'} - ${urlStr} (${reason})`);
  } catch (e) {
    console.log(`ERROR - ${urlStr} (${e.message})`);
  }
}

async function main() {
  for (const urlStr of urls) {
    await checkUrl(urlStr);
  }
}

main();
